### ISSUE TITLE
False positive boxed_local

### ISSUE THREAD

**Issue by dwijnand** [rust-lang-nursery/rust-clippy#2950]
Given

```rust
trait FnBox<A, R> {
    fn call_box(self: Box<Self>, a: A) -> R;
}

impl<A, R, F: FnOnce(A) -> R> FnBox<A, R> for F {
    fn call_box(self: Box<F>, a: A) -> R {
        (*self)(a)
    }
}
```

Clippy warns:

```rust
warning: local variable doesn't need to be boxed here
 --> src/lib.rs:6:17
  |
6 |     fn call_box(self: Box<F>, a: A) -> R {
  |                 ^^^^
  |
  = note: #[warn(boxed_local)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#boxed_local
```

Quoting @oli-obk:

> Looks like a clippy bug. we should not lint trait methods

(https://github.com/rust-lang/cargo/pull/5768#discussion_r204291733)

**Comment by dwijnand** [comment]
Duplicate of #1123
